elastic/beats
# [Journald] Restart journalctl if it exits unexpectedly

## Proposed commit message

If journalctl exits unexpectedly the journald input will restart it and set the cursor to the last know position. Any error/non zero return code is logged at level error.


## Checklist

## Disruptive User Impact


## Author's Checklist

- Test if this PR fixes https://github.com/elastic/beats/issues/32782

## How to test this PR locally

1. Build Filebeat from this PR and start it with the following configuration file:
<details><summary>filebeat.yml</summary>
<p>

```yaml
filebeat.inputs:
  - type: journald
    id: my-unique-input-id
    enabled: true
    syslog_identifiers:
      - "potato"

output:
  console:
    pretty: true
```

</p>
</details> 


2. Publish some events with the syslog identfier `potato`

```
echo "$(date '+%Y-%m-%d %T') - foo"| systemd-cat -t potato
```

3. Wait to see the output in the console
4. Kill journalctl

```
killall journalctl
```


5. Wait journalctl to restart, look for a log entry like this:

```json
{
  "log.level": "info",
  "@timestamp": "2024-08-20T12:41:29.682-0400",
  "log.logger": "input.journald.reader.journalctl-runner",
  "log.origin": {
    "function": "github.com/elastic/beats/v7/filebeat/input/journald/pkg/journalctl.Factory",
    "file.name": "journalctl/journalctl.go",
    "file.line": 120
  },
  "message": "journalctl started with PID 3119456",
  "service.name": "filebeat",
  "id": "filestream-input-id",
  "input_source": "LOCAL_SYSTEM_JOURNAL",
  "path": "LOCAL_SYSTEM_JOURNAL",
  "ecs.version": "1.6.0"
}
```

6. Publish a few more messages and ensure they appear in the console

```
echo "$(date '+%Y-%m-%d %T') - foo"| systemd-cat -t potato
```


## Related issues

- Closes https://github.com/elastic/beats/issues/39352

## Tests
I run a test overnight where I used a [mock](https://github.com/elastic/beats/blob/d5cb6fce5e607c69b38fb8d7717e99d5c98e325b/filebeat/tests/mockjournalctl/main.go) to simulate constant failures and restarts of journalclt and monitored the host and Filebeat process to ensure there weren't any goroutine/resouce leaks. The screenshot below shows the CPU and memory usage from Filebeat as well as counters for the log entries stating journalctl crashed and was restarted.

![2024-09-10_09-14](https://github.com/user-attachments/assets/cd864316-d861-477c-ae80-57ebeb9b8a4a)


~~## Use cases~~
~~## Screenshots~~
~~## Logs~~
